File: go/json_source_locator.go
// Package jsonstructure provides validators for JSON Structure schemas and instances.
package jsonstructure

import (
	"strconv"
	"strings"
)

// JsonSourceLocator tracks line and column positions in a JSON document
// and maps JSON Pointer paths to source locations.
type JsonSourceLocator struct {
	jsonText    string
	lineOffsets []int
}

// NewJsonSourceLocator creates a new source locator for the given JSON text.
func NewJsonSourceLocator(jsonText string) *JsonSourceLocator {
	locator := &JsonSourceLocator{
		jsonText:    jsonText,
		lineOffsets: buildLineOffsets(jsonText),
	}
	return locator
}

// buildLineOffsets builds an array of line start offsets for efficient line/column lookup.
func buildLineOffsets(text string) []int {
	offsets := []int{0} // First line starts at offset 0
	for i, char := range text {
		if char == '\n' {
			offsets = append(offsets, i+1)
		}
	}
	return offsets
}

// GetLocation returns the source location for a JSON Pointer path.
func (l *JsonSourceLocator) GetLocation(path string) JsonLocation {
	if path == "" || l.jsonText == "" {
		return UnknownLocation()
	}

	// Parse the JSON Pointer path
	segments := l.parseJSONPointer(path)

	// Navigate through the JSON text to find the location
	return l.findLocationInText(segments)
}

// parseJSONPointer parses a JSON Pointer path into segments.
// Handles both standard JSON Pointer format (/a/0) and bracket notation (/a[0]).
func (l *JsonSourceLocator) parseJSONPointer(path string) []string {
	// Remove leading # if present (JSON Pointer fragment identifier)
	path = strings.TrimPrefix(path, "#")

	// Handle empty path or just "/"
	if path == "" || path == "/" {
		return []string{}
	}

	// Split by / and unescape segments, handling bracket notation
	var segments []string
	for _, segment := range strings.Split(path, "/") {
		if segment == "" {
			continue
		}
		// Unescape JSON Pointer tokens
		segment = strings.ReplaceAll(segment, "~1", "/")
		segment = strings.ReplaceAll(segment, "~0", "~")
		
		// Handle bracket notation (e.g., "required[0]" -> "required", "0")
		if bracketIdx := strings.Index(segment, "["); bracketIdx > 0 {
			// Extract the property name before the bracket
			propName := segment[:bracketIdx]
			segments = append(segments, propName)
			
			// Extract all bracket indices
			rest := segment[bracketIdx:]
			for len(rest) > 0 && rest[0] == '[' {
				closeIdx := strings.Index(rest, "]")
				if closeIdx < 0 {
					break
				}
				indexStr := rest[1:closeIdx]
				segments = append(segments, indexStr)
				rest = rest[closeIdx+1:]
			}
		} else {
			segments = append(segments, segment)
		}
	}

	return segments
}

// offsetToLocation converts a character offset to line/column position.
func (l *JsonSourceLocator) offsetToLocation(offset int) JsonLocation {
	if offset < 0 || offset > len(l.jsonText) {
		return UnknownLocation()
	}

	// Binary search for the line
	line := 0
	left, right := 0, len(l.lineOffsets)-1
	for left <= right {
		mid := (left + right) / 2
		if l.lineOffsets[mid] <= offset {
			line = mid
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	// Column is offset from line start (1-based)
	column := offset - l.lineOffsets[line] + 1
	return JsonLocation{Line: line + 1, Column: column} // 1-based line numbers
}

// findLocationInText navigates through JSON text to find the location of a path.
func (l *JsonSourceLocator) findLocationInText(segments []string) JsonLocation {
	text := l.jsonText
	offset := 0

	// Skip initial whitespace
	offset = l.skipWhitespace(text, offset)

	if offset >= len(text) {
		return UnknownLocation()
	}

	// If no segments, return the start of the document
	if len(segments) == 0 {
		return l.offsetToLocation(offset)
	}

	// Navigate through each segment
	for _, segment := range segments {
		offset = l.skipWhitespace(text, offset)

		if offset >= len(text) {
			return UnknownLocation()
		}

		char := text[offset]

		if char == '{' {
			// Object - find the property
			offset = l.findObjectProperty(text, offset, segment)
			if offset < 0 {
				return UnknownLocation()
			}
		} else if char == '[' {
			// Array - find the index
			index, err := strconv.Atoi(segment)
			if err != nil {
				return UnknownLocation()
			}
			offset = l.findArrayElement(text, offset, index)
			if offset < 0 {
				return UnknownLocation()
			}
		} else {
			// Not an object or array, can't navigate further
			return UnknownLocation()
		}
	}

	return l.offsetToLocation(offset)
}

// skipWhitespace skips whitespace characters.
func (l *JsonSourceLocator) skipWhitespace(text string, offset int) int {
	for offset < len(text) {
		c := text[offset]
		if c != ' ' && c != '\t' && c != '\n' && c != '\r' {
			break
		}
		offset++
	}
	return offset
}

// skipString skips a JSON string value and returns the offset after the closing quote.
func (l *JsonSourceLocator) skipString(text string, offset int) int {
	if offset >= len(text) || text[offset] != '"' {
		return offset
	}

	offset++ // Skip opening quote
	for offset < len(text) {
		char := text[offset]
		if char == '\\' {
			offset += 2 // Skip escape sequence
		} else if char == '"' {
			return offset + 1 // Return position after closing quote
		} else {
			offset++
		}
	}
	return offset
}

// skipValue skips a JSON value and returns the offset after it.
func (l *JsonSourceLocator) skipValue(text string, offset int) int {
	offset = l.skipWhitespace(text, offset)

	if offset >= len(text) {
		return offset
	}

	char := text[offset]

	switch {
	case char == '"':
		return l.skipString(text, offset)
	case char == '{':
		return l.skipObject(text, offset)
	case char == '[':
		return l.skipArray(text, offset)
	case char == 't':
		if offset+4 <= len(text) && text[offset:offset+4] == "true" {
			return offset + 4
		}
	case char == 'f':
		if offset+5 <= len(text) && text[offset:offset+5] == "false" {
			return offset + 5
		}
	case char == 'n':
		if offset+4 <= len(text) && text[offset:offset+4] == "null" {
			return offset + 4
		}
	case char == '-' || (char >= '0' && char <= '9'):
		// Number
		for offset < len(text) {
			c := text[offset]
			if c != '-' && c != '+' && c != '.' && c != 'e' && c != 'E' && (c < '0' || c > '9') {
				break
			}
			offset++
		}
		return offset
	}

	return offset
}

// skipObject skips an entire JSON object.
func (l *JsonSourceLocator) skipObject(text string, offset int) int {
	if offset >= len(text) || text[offset] != '{' {
		return offset
	}

	offset++ // Skip opening brace
	depth := 1

	for offset < len(text) && depth > 0 {
		offset = l.skipWhitespace(text, offset)
		if offset >= len(text) {
			break
		}

		char := text[offset]
		switch char {
		case '{':
			depth++
			offset++
		case '}':
			depth--
			offset++
		case '"':
			offset = l.skipString(text, offset)
		default:
			offset++
		}
	}

	return offset
}

// skipArray skips an entire JSON array.
func (l *JsonSourceLocator) skipArray(text string, offset int) int {
	if offset >= len(text) || text[offset] != '[' {
		return offset
	}

	offset++ // Skip opening bracket
	depth := 1

	for offset < len(text) && depth > 0 {
		offset = l.skipWhitespace(text, offset)
		if offset >= len(text) {
			break
		}

		char := text[offset]
		switch char {
		case '[':
			depth++
			offset++
		case ']':
			depth--
			offset++
		case '{':
			offset = l.skipObject(text, offset)
		case '"':
			offset = l.skipString(text, offset)
		default:
			offset++
		}
	}

	return offset
}

// findObjectProperty finds a property in an object and returns the offset of its value.
func (l *JsonSourceLocator) findObjectProperty(text string, offset int, propertyName string) int {
	if offset >= len(text) || text[offset] != '{' {
		return -1
	}

	offset++ // Skip opening brace

	for offset < len(text) {
		offset = l.skipWhitespace(text, offset)

		if offset >= len(text) {
			return -1
		}

		if text[offset] == '}' {
			return -1 // End of object, property not found
		}

		// Skip comma if present
		if text[offset] == ',' {
			offset++
			offset = l.skipWhitespace(text, offset)
		}

		// Expect a property name (string)
		if offset >= len(text) || text[offset] != '"' {
			return -1
		}

		// Parse the property name
		nameStart := offset + 1
		offset = l.skipString(text, offset)
		nameEnd := offset - 1 // Don't include closing quote
		currentName := text[nameStart:nameEnd]

		// Skip whitespace and colon
		offset = l.skipWhitespace(text, offset)
		if offset >= len(text) || text[offset] != ':' {
			return -1
		}
		offset++
		offset = l.skipWhitespace(text, offset)

		if currentName == propertyName {
			return offset // Return offset of the value
		}

		// Skip this value
		offset = l.skipValue(text, offset)
	}

	return -1
}

// findArrayElement finds an array element by index and returns the offset of its value.
func (l *JsonSourceLocator) findArrayElement(text string, offset int, index int) int {
	if offset >= len(text) || text[offset] != '[' {
		return -1
	}

	offset++ // Skip opening bracket
	currentIndex := 0

	for offset < len(text) {
		offset = l.skipWhitespace(text, offset)

		if offset >= len(text) {
			return -1
		}

		if text[offset] == ']' {
			return -1 // End of array, index not found
		}

		// Skip comma if present
		if text[offset] == ',' {
			offset++
			offset = l.skipWhitespace(text, offset)
		}

		if currentIndex == index {
			return offset // Return offset of the element
		}

		// Skip this value
		offset = l.skipValue(text, offset)
		currentIndex++
	}

	return -1
}

File: typescript/src/json-source-locator.ts
/**
 * Locates the source position (line and column) of a JSON Pointer path within JSON text.
 */

import { JsonLocation, UNKNOWN_LOCATION } from './types';

/**
 * Parses a JSON Pointer path and returns an array of segments.
 * Handles both standard JSON Pointer format (/foo/bar) and 
 * the notation used in validation paths (#/foo/bar[0]).
 */
function parsePath(path: string): string[] {
  if (!path) {
    return [];
  }

  // Remove leading '#' if present (JSON Pointer anchor)
  if (path.startsWith('#')) {
    path = path.slice(1);
  }

  // Remove leading slash
  if (path.startsWith('/')) {
    path = path.slice(1);
  }

  if (!path) {
    return [];
  }

  // Split and unescape JSON Pointer tokens
  // Also handle array index notation like [0] or [1]
  const segments: string[] = [];
  for (const segment of path.split('/')) {
    // Unescape JSON Pointer tokens
    const unescaped = segment.replace(/~1/g, '/').replace(/~0/g, '~');
    
    // Check for array index notation like "required[0]"
    const bracketMatch = unescaped.match(/^(.+?)\[(\d+)\]$/);
    if (bracketMatch) {
      // Split into property name and array index
      segments.push(bracketMatch[1]);
      segments.push(bracketMatch[2]);
    } else {
      segments.push(unescaped);
    }
  }
  
  return segments;
}

/**
 * JSON Source Locator class that finds line/column positions from JSON Pointer paths.
 */
export class JsonSourceLocator {
  private readonly json: string;
  private readonly locationCache: Map<string, JsonLocation> = new Map();

  constructor(json: string) {
    this.json = json;
  }

  /**
   * Gets the source location for a JSON Pointer path.
   */
  getLocation(path: string): JsonLocation {
    if (!path) {
      return this.getRootLocation();
    }

    const cached = this.locationCache.get(path);
    if (cached) {
      return cached;
    }

    const location = this.findLocationForPath(path);
    this.locationCache.set(path, location);
    return location;
  }

  private getRootLocation(): JsonLocation {
    let line = 1;
    let column = 1;

    for (const c of this.json) {
      if (!/\s/.test(c)) {
        return { line, column };
      }

      if (c === '\n') {
        line++;
        column = 1;
      } else if (c !== '\r') {
        column++;
      }
    }

    return { line: 1, column: 1 };
  }

  private findLocationForPath(path: string): JsonLocation {
    try {
      const segments = parsePath(path);
      if (segments.length === 0) {
        return this.getRootLocation();
      }

      // Use a simple approach: parse the JSON and track positions
      let position = 0;
      let line = 1;
      let column = 1;

      const updatePosition = (chars: number) => {
        for (let i = 0; i < chars && position + i < this.json.length; i++) {
          if (this.json[position + i] === '\n') {
            line++;
            column = 1;
          } else if (this.json[position + i] !== '\r') {
            column++;
          }
        }
        position += chars;
      };

      const skipWhitespace = () => {
        while (position < this.json.length && /\s/.test(this.json[position])) {
          updatePosition(1);
        }
      };

      const readString = (): string => {
        if (this.json[position] !== '"') {
          throw new Error('Expected string');
        }
        updatePosition(1);
        let result = '';
        while (position < this.json.length && this.json[position] !== '"') {
          if (this.json[position] === '\\') {
            updatePosition(1);
            if (position < this.json.length) {
              result += this.json[position];
              updatePosition(1);
            }
          } else {
            result += this.json[position];
            updatePosition(1);
          }
        }
        updatePosition(1); // closing quote
        return result;
      };

      const skipValue = () => {
        skipWhitespace();
        const c = this.json[position];
        if (c === '"') {
          readString();
        } else if (c === '{') {
          let depth = 1;
          updatePosition(1);
          while (position < this.json.length && depth > 0) {
            if (this.json[position] === '{') {
              depth++;
              updatePosition(1);
            } else if (this.json[position] === '}') {
              depth--;
              updatePosition(1);
            } else if (this.json[position] === '"') {
              readString();
            } else {
              updatePosition(1);
            }
          }
        } else if (c === '[') {
          let depth = 1;
          updatePosition(1);
          while (position < this.json.length && depth > 0) {
            if (this.json[position] === '[') {
              depth++;
              updatePosition(1);
            } else if (this.json[position] === ']') {
              depth--;
              updatePosition(1);
            } else if (this.json[position] === '"') {
              readString();
            } else {
              updatePosition(1);
            }
          }
        } else {
          // number, boolean, null
          while (position < this.json.length && /[^,\]\}]/.test(this.json[position])) {
            updatePosition(1);
          }
        }
      };

      // Navigate to the path
      let currentSegmentIndex = 0;

      skipWhitespace();

      while (currentSegmentIndex < segments.length && position < this.json.length) {
        const segment = segments[currentSegmentIndex];
        const isArrayIndex = /^\d+$/.test(segment);

        skipWhitespace();

        if (isArrayIndex) {
          // Navigate into array
          if (this.json[position] !== '[') {
            return UNKNOWN_LOCATION;
          }
          updatePosition(1);
          skipWhitespace();

          const targetIndex = parseInt(segment, 10);
          let currentIndex = 0;

          while (position < this.json.length && this.json[position] !== ']') {
            if (currentIndex === targetIndex) {
              if (currentSegmentIndex === segments.length - 1) {
                // Found it!
                return { line, column };
              }
              currentSegmentIndex++;
              break;
            }

            skipValue();
            skipWhitespace();

            if (this.json[position] === ',') {
              updatePosition(1);
              skipWhitespace();
              currentIndex++;
            } else {
              break;
            }
          }

          if (currentIndex !== targetIndex && currentSegmentIndex !== segments.length) {
            return UNKNOWN_LOCATION;
          }
        } else {
          // Navigate into object
          if (this.json[position] !== '{') {
            return UNKNOWN_LOCATION;
          }
          updatePosition(1);
          skipWhitespace();

          let found = false;

          while (position < this.json.length && this.json[position] !== '}') {
            // Read property name
            const propName = readString();
            skipWhitespace();

            if (this.json[position] !== ':') {
              return UNKNOWN_LOCATION;
            }
            updatePosition(1);
            skipWhitespace();

            if (propName === segment) {
              if (currentSegmentIndex === segments.length - 1) {
                // Found it!
                return { line, column };
              }
              currentSegmentIndex++;
              found = true;
              break;
            }

            // Skip this value
            skipValue();
            skipWhitespace();

            if (this.json[position] === ',') {
              updatePosition(1);
              skipWhitespace();
            }
          }

          if (!found) {
            return UNKNOWN_LOCATION;
          }
        }
      }

      return UNKNOWN_LOCATION;
    } catch {
      return UNKNOWN_LOCATION;
    }
  }
}
